Mathlib/LinearAlgebra/Matrix/BilinearForm.lean

```diff
@@ -186,87 +271,213 @@ theorem Matrix.toBilin'_stdBasis (M : Matrix n n R₂) (i j : n) :
   Matrix.toBilin'Aux_stdBasis M i j
 #align matrix.to_bilin'_std_basis Matrix.toBilin'_stdBasis
 
+@[simp]
+theorem BilinForm.toMatrix'_symm' :
+    (BilinForm.toMatrix''.symm : Matrix n n N₂ ≃ₗ[R₂] _) = Matrix.toBilin'' :=
+  rfl
+
 @[simp]
 theorem BilinForm.toMatrix'_symm :
     (BilinForm.toMatrix'.symm : Matrix n n R₂ ≃ₗ[R₂] _) = Matrix.toBilin' :=
   rfl
 #align bilin_form.to_matrix'_symm BilinForm.toMatrix'_symm
 
+@[simp]
+theorem Matrix.toBilin'_symm' :
+    (Matrix.toBilin''.symm : _ ≃ₗ[R₂] Matrix n n N₂) = BilinForm.toMatrix'' :=
+  BilinForm.toMatrix''.symm_symm
+
 @[simp]
 theorem Matrix.toBilin'_symm :
     (Matrix.toBilin'.symm : _ ≃ₗ[R₂] Matrix n n R₂) = BilinForm.toMatrix' :=
   BilinForm.toMatrix'.symm_symm
 #align matrix.to_bilin'_symm Matrix.toBilin'_symm
 
+@[simp]
+theorem Matrix.toBilin'_toMatrix'' (B : (n → R₂) →ₗ[R₂] (n → R₂) →ₗ[R₂] N₂) :
+    Matrix.toBilin'' (BilinForm.toMatrix'' B) = B :=
+  Matrix.toBilin''.apply_symm_apply B
+
 @[simp]
 theorem Matrix.toBilin'_toMatrix' (B : BilinForm R₂ (n → R₂)) :
     Matrix.toBilin' (BilinForm.toMatrix' B) = B :=
   Matrix.toBilin'.apply_symm_apply B
 #align matrix.to_bilin'_to_matrix' Matrix.toBilin'_toMatrix'
 
+@[simp]
+theorem BilinForm.toMatrix'_toBilin'' (M : Matrix n n N₂) :
+    BilinForm.toMatrix'' (Matrix.toBilin'' (R₂ := R₂) M) = M :=
+  BilinForm.toMatrix''.apply_symm_apply M
+
 @[simp]
 theorem BilinForm.toMatrix'_toBilin' (M : Matrix n n R₂) :
     BilinForm.toMatrix' (Matrix.toBilin' M) = M :=
   BilinForm.toMatrix'.apply_symm_apply M
 #align bilin_form.to_matrix'_to_bilin' BilinForm.toMatrix'_toBilin'
 
+@[simp]
+theorem BilinForm.toMatrix'_apply' (B : (n → R₂) →ₗ[R₂] (n → R₂) →ₗ[R₂] N₂) (i j : n) :
+    BilinForm.toMatrix'' B i j = B (stdBasis R₂ (fun _ => R₂) i 1)
+    (stdBasis R₂ (fun _ => R₂) j 1) :=
+  rfl
+
 @[simp]
 theorem BilinForm.toMatrix'_apply (B : BilinForm R₂ (n → R₂)) (i j : n) :
     BilinForm.toMatrix' B i j = B (stdBasis R₂ (fun _ => R₂) i 1) (stdBasis R₂ (fun _ => R₂) j 1) :=
   rfl
 #align bilin_form.to_matrix'_apply BilinForm.toMatrix'_apply
 
--- Porting note: dot notation for bundled maps doesn't work in the rest of this section
+@[default_instance 100]
+instance SMatrixLeftMul {m : Type*} :
+    HSMul (Matrix m n R₂) (Matrix n o N₂) (Matrix m o N₂) where
+  hSMul M₁ M₂ := fun i k => ∑ j, M₁ i j • M₂ j k
+
+@[default_instance 100]
+instance SMatrixRightMul {m : Type*} :
+    HSMul (Matrix m n N₂) (Matrix n o R₂) (Matrix m o N₂) where
+  hSMul M₁ M₂ := fun i k => ∑ j, M₂ j k • M₁ i j
```

Comment:
I think these instances are pretty bad if `N₂ = R₂`, because the end up both applying but being unequal to each other. I think generally it's impossible to reasonably use `HSMul` without introducing diamonds.

Reply:
Multiplication by a matrix of scalars doesn't seem entirely natural to me - I can't think of anywhere else where this happens - but I wasn't sure how else to generalise this bit.

Reply:
@eric-wieser would something like this be better?
```
def SMatrixLeftMul {m : Type*} :
    HSMul (Matrix m n R₂) (Matrix n o N₂) (Matrix m o N₂) where
  hSMul M₁ M₂ := fun i k => ∑ j, M₁ i j • M₂ j k

def SMatrixRightMul {m : Type*} :
    HSMul (Matrix m n N₂) (Matrix n o R₂) (Matrix m o N₂) where
  hSMul M₁ M₂ := fun i k => ∑ j, M₂ j k • M₁ i j

notation M₁ "•ₗ" M₂ => SMatrixLeftMul.hSMul M₁ M₂

notation M₂ "•ᵣ" M₁ => SMatrixRightMul.hSMul M₂ M₁
```

Reply:
However, Lean then interprets:
```
variable (M : Matrix n n N₂)

#check (1 : Matrix n n R₂) •ₗ M = M
```
as `(1 : Matrix n n R₂) •ₗ (M = M)` rather than `((1 : Matrix n n R₂) •ₗ M) = M` - so I need some way to set the priority?

Reply:
This appears to work:
```
def SMatrixLeftMul  {m : Type*} :
    HSMul (Matrix m n R₂) (Matrix n o N₂) (Matrix m o N₂) where
  hSMul M₁ M₂ := fun i k => ∑ j, M₁ i j • M₂ j k

def SMatrixRightMul {m : Type*} :
    HSMul (Matrix m n N₂) (Matrix n o R₂) (Matrix m o N₂) where
  hSMul M₁ M₂ := fun i k => ∑ j, M₂ j k • M₁ i j

local notation:100 M₁ "•ₗ" M₂:100 => SMatrixLeftMul.hSMul M₁ M₂

local notation:100 M₂ "•ᵣ" M₁:100 => SMatrixRightMul.hSMul M₂ M₁
```